keps/sig-node/20190726-arbitrary-pull-secrets.md

```diff
@@ -0,0 +1,98 @@
+---
+title: Arbitrary image pull secrets
+authors:
+  - "@sashayakovtseva"
+owning-sig: sig-node
+creation-date: 2019-07-26
+status: provisional
+---
+
+# Arbitrary image pull secrets
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+  - [Implementation Details](#implementation-details)
+- [Alternatives ](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Currently in order to pull an image from a private registry one can use `imagePullSecrets` (as described
+[here](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/)). 
+However, it is tightly integrated with Docker and there is no other way to support different on-prem
+registries or different types of authorization (e.g. private tokens).
+
+This proposal suggests changes that will fully enable arbitrary private image registry support.
+
+## Motivation
+
+Kubernetes introduced CRI making it possible to integrate any container runtime with it.
+CRI contains `ImageService` that is responsible for any image-related operations performed by kubelet. 
+
+Whenever a new container is created and base image is not present on node, `PullImage` method of `ImageService`
+is called. `PullImageRequest` contains `AuthConfig`, so no changes should be make in CRI to enable support
+for private image pulls.   
+
+It is common to have private container images or on-prem image registries when working with Kubernetes. To use
+those, one should either perform node configuration or leverage `imagePullSecrets` for pod that will be passed
+to the CRI implementation in `AuthConfig`.
+ 
+However, even though Kubernetes has CRI, `imagePullSecrets` are respected only when the secret type is
+`kubernetes.io/dockerconfigjson` or `kubernetes.io/dockercfg`. This is not aligned with CRI concept since CRI should
+enable support for arbitrary container image registries (via `ImageService`) and arbitrary container runtimes (via
+`RuntimeService`). 
+
+### Goals
+
+- Introduce a new secret type, e.g. `kubernetes.io/regcredentials`.
+It should contain data that can be mapped to the CRI's `AuthConfig` subset of fields directly.
```

Comment:
I'm not sure whether we want to define the schema based on `AuthConfig`.

CRI is a Kubelet internal API, and we can't guarantee that we won't change or extend `AuthConfig` in the future. If we changed it, how do we handle the version skew?

Reply:
Another option would be to pass pull secrets as map[string]string or smth and let CRI decide what to do, but I am not sure that is better than `AuthConfig` mapping.

Reply:
@sashayakovtseva I'm a bit confused, does the `AuthConfig` object meet your requirement?

If it does, it should be equivalent to today's docker secret https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#registry-secret-existing-credentials. Kubelet just converts docker secret into `AuthConfig`.

If `AuthConfig` works for you, you just need to fill in the value into corresponding fields in the docker secret, and then the docker secret should also work for you.

Reply:
@Random-Liu `AuthConfig` can be used. We are using access tokens to pull private images, so `AuthConfig.IdentityToken` is the proper field to use IMO. 

And yes, the workaround you suggest *partially* works. Here are the issues that should be addressed:
1. Forcing any CRI to use secret of type `kubernetes.io/dockerconfigjson` doesn't match with idea of abstracting runtime with CRI. It also requires username which we don't need because of access tokens instead of username/password pair.
2.  I could not find where `AuthConfig.ServerAddress` is set on kubelet side. Without that it is not possible to use on-prem image registries which we want to leverage.

That is why even though with a workaround it is possible to pull a private image from default registry, I still believe this should be changed on kubelet side to fully enable runtime abstraction.

Reply:
It is in the keyring https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L66,
and kubelet does pass it to CRI https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_image.go#L63.

I'm totally in to a more generic credential format in the Kubernetes api, instead of `kubernetes.io/dockerconfigjson`. However, we need a use case with more representativeness to define the new API, so we are waiting for a use case that can't be covered by today's `kubernetes.io/dockerconfigjson`. :)

Reply:
I think @Random-Liu addressed the answer to point (2), I am missing where (1) requires a username.  It's possible I am missing something.  Either way, I am in favor of supporting a generic credential format and/or secret type.

Reply:
@Random-Liu Regarding point 2. If I am not missing anything, this is where the issue with `ServerAddress` lies. I tried to follow pull image flow to check where `ServerAddress` is really set:
- [make a keyring based on secrets](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_image.go#L38)
- [unmarshal passed secret into docker config](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/secrets/secrets.go#L34)
- [resulting struct does not have `ServerAddress`](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/config.go#L50)
- [during creds add no `ServerAddress` is set too](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L82)
- [during lookup either](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L243)

So if above is correct, then [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_image.go#L63) assignment is doing nothing.

Please, correct me if I am wrong and point to the line where `ServerAddress` is set based on the passed secret.

Reply:
@sashayakovtseva Hm, I think you are right!

However, I don't understand why we don't support `ServerAddress`, `IdentityToken` and `RegistryToken` in the first place.

Reply:
@Random-Liu IMO (1) should be a temporarily fix unless Kubernetes wishes to fully support only Docker containers even though there will be other CRI implementations. 
I understand that _technically_ this will work, but that is really awkward to ask e.g. Singularity users to create Docker secret to pull private images...

Reply:
I like 2) in the long term, but not sure whether we can define the new format correctly at the moment. Thus I said I prefer 1) for now.

If @kubernetes/sig-auth-api-reviews thinks that the `AuthConfig` format is sufficient and generic enough for Kubernetes, I'm totally fine with 2). :)

Reply:
BTW, the `ServerAddress` should be supported to handle the case mentioned in https://github.com/containerd/cri/pull/1227#discussion_r312253579

Reply:
+1 to @Random-Liu's suggestion. 

I'd like to API to look more generic as well, but given what we have today, it's preferred if we can easily extend the existing API to support your case. If there are other use cases that cannot be easily satisfied, we can introduce a new type.